### PR TITLE
Improvements to the Helix BootstrapUpgrade tool

### PR DESCRIPTION
**Pull request by pnarayanan**
- Ensure that InstanceConfig does not get updated unless required,
during upgrades.
- Ensure reconciliation of changes to partition/replica capacity of
already added partitions.

### CODE REVIEW

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 1]
```diff
@@ -271,36 +271,101 @@ private void addNewDataNodes() {
...
         DataNodeId node = replicaId.getDataNodeId();
         String instanceName = getInstanceName(node);
         InstanceConfig instanceConfig = dcAdmin.getInstanceConfig(clusterName, instanceName);
-        List<String> currentSealedPartitions = instanceConfig.getRecord().getListField(ClusterMapUtils.SEALED_STR);
-        List<String> newSealedPartitionsList = new ArrayList<>(currentSealedPartitions);
-        if (isSealed && !currentSealedPartitions.contains(partitionName)) {
-          newSealedPartitionsList.add(partitionName);
-        } else if (!isSealed && currentSealedPartitions.contains(partitionName)) {
-          newSealedPartitionsList.remove(partitionName);
+        boolean shouldSetInstanceConfig = false;
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: updateInstanceConfig might be apt name

Comment (pnarayanan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: I actually prefer it the way it is, as we are setting the instanceConfig later. The "update" prefix is generally being used for method names in this class.

**Comment by pnarayanan** [comment]
@nsivabalan @cgtz Built and formatted. Please take a look.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 2]
```diff
@@ -267,40 +267,106 @@ private void addNewDataNodes() {
...
+   * @param instanceConfig the {@link InstanceConfig} of the node.
+   * @param mountPath the mount path of the replica.
+   * @return true, if the {@link InstanceConfig} was updated by this method; false otherwise.
+   */
+  private boolean updateReplicaCapacityIfRequired(String partitionName, InstanceConfig instanceConfig, String mountPath,
+      long actualReplicaCapacity) {
+    boolean instanceConfigUpdated = false;
+    Map<String, String> diskInfo = instanceConfig.getRecord().getMapField(mountPath);
+    String currentReplicasStr = diskInfo.get(ClusterMapUtils.REPLICAS_STR);
+    String newReplicaStr = "";
```
Comment (cgtz) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Could you use a `StringBuilder` here?

File: ambry-tools/src/integration-test/java/com.github.ambry.clustermap/HelixBootstrapUpgradeToolTest.java [hunk 3]
```diff
@@ -135,11 +135,12 @@ public void testEverything() throws Exception {
     testHardwareLayout.addNewDataNodes(numNewNodes);
     testPartitionLayout.addNewPartitions(numNewPartitions);
 
-    // Next, mark all previous partitions as READ_ONLY
+    // Next, mark all previous partitions as READ_ONLY, and change their replica capacities.
     for (PartitionId partitionId : testPartitionLayout.getPartitionLayout().getPartitions()) {
       if (partitionIdsBeforeAddition.contains(((Partition) partitionId).getId())) {
         Partition partition = (Partition) partitionId;
         partition.partitionState = PartitionState.READ_ONLY;
+        partition.replicaCapacityInBytes += 1;
```
Comment (cgtz) [inline ambry-tools/src/integration-test/java/com.github.ambry.clustermap/HelixBootstrapUpgradeToolTest.java]: Why does the replicaCapacity only increase by one byte?

Comment (pnarayanan) [inline ambry-tools/src/integration-test/java/com.github.ambry.clustermap/HelixBootstrapUpgradeToolTest.java]: This is testing the case where replica capacity changes between upgrades, so adding 1 to change it.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Done (probably not a big deal though as this does not run within a service).

**Comment by pnarayanan** [comment]
Updated, built and tested.

File: ambry-tools/src/integration-test/java/com.github.ambry.clustermap/HelixBootstrapUpgradeToolTest.java (hunk 3 above)
Comment (cgtz) [inline ambry-tools/src/integration-test/java/com.github.ambry.clustermap/HelixBootstrapUpgradeToolTest.java]: Ok, that makes sense.

**Pull request merged**